compiler/src/dotty/tools/dotc/inlines/InlineReducer.scala

```diff
@@ -368,29 +413,30 @@ class InlineReducer(inliner: Inliner)(using Context):
 
       if (!isImplicit) caseBindingMap += ((NoSymbol, scrutineeBinding))
       val gadtCtx = ctx.fresh.setFreshGADTBounds.addMode(Mode.GadtConstraintInference)
-      if (reducePattern(caseBindingMap, scrutineeSym.termRef, cdef.pat)(using gadtCtx)) {
-        val (caseBindings, from, to) = substBindings(caseBindingMap.toList, mutable.ListBuffer(), Nil, Nil)
-        val (guardOK, canReduceGuard) =
-          if cdef.guard.isEmpty then (true, true)
-          else typer.typed(cdef.guard.subst(from, to), defn.BooleanType) match {
-            case ConstantValue(v: Boolean) => (v, true)
-            case _ => (false, false)
-          }
-        if guardOK then Some((caseBindings.map(_.subst(from, to)), cdef.body.subst(from, to), canReduceGuard))
-        else if canReduceGuard then None
-        else Some((caseBindings.map(_.subst(from, to)), cdef.body.subst(from, to), canReduceGuard))
-      }
-      else None
+      reducePattern(caseBindingMap,scrutineeSym.termRef,cdef.pat)(using gadtCtx) match
+        case Some(msg) => Left(msg)
+        case None => {
+          val (caseBindings, from, to) = substBindings(caseBindingMap.toList, mutable.ListBuffer(), Nil, Nil)
+          val (guardOK, canReduceGuard) =
+            if cdef.guard.isEmpty then (true, true)
+            else typer.typed(cdef.guard.subst(from, to), defn.BooleanType) match {
+              case ConstantValue(v: Boolean) => (v, true)
+              case _ => (false, false)
+            }
+          if guardOK then Right((caseBindings.map(_.subst(from, to)), cdef.body.subst(from, to), canReduceGuard))
+          else if canReduceGuard then Left(None)
+          else Right((caseBindings.map(_.subst(from, to)), cdef.body.subst(from, to), canReduceGuard))
+        }
     }
 
-    def recur(cases: List[CaseDef]): MatchRedux = cases match {
-      case Nil => None
+    def recur(cases: List[CaseDef], theLastMsg: Option[Message] = None): MatchRedux = cases match
+      case Nil => Left(theLastMsg)
       case cdef :: cases1 =>
         reduceCase(cdef) match
-          case None => recur(cases1)
-          case r @ Some((caseBindings, rhs, canReduceGuard)) if canReduceGuard => Some((caseBindings, rhs))
-          case _ => None
-    }
+          case Left(None) => recur(cases1, theLastMsg)
+          case Left(msg @Some(_)) => recur(cases1, msg)
```

Comment:
Should it fail fast when a case results in missing implicit instance rather than continuing remaining cases?